### PR TITLE
feat: add login page

### PR DESCRIPTION
**Pull request by kxrt**
- Create login page on route '/login'
- Implement regex check on email to validate inputs better
- Add 'Login' object to Layout bar
- Implement usage of snackbars on successful/unsuccesful login attempt
- Set up basic styling, using UI colors from Trello
- Split half-and-half with placeholder image and form
  - The image vanishes on smaller devices to save space, which can be changed later
- Add FoodBank logo
- Implement useMutation to set up login form
- Wrap app in QueryProvider
- Set up file for login api call, to be merged with other api calls when ready

<p float="left" align="middle">
  Login_Web
  Login_Tablet
  Login_Mobile
</p>

### CODE REVIEW

File: pages/login/index.tsx [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+              src='https://foodbank.sg/wp-content/uploads/2020/07/IMG_9279-1024x768.jpg'
+              alt='Food Bank'
+              backgroundPosition='center'
+              height='100vh'
+              fit='cover'
+            />
+          </Box>
+          <Box
+            marginBlock={[2, 0, 0, 0]}
+            p={8}
```
Comment (MorningLit) [inline pages/login/index.tsx:64]: height overflow because of this padding

File: pages/login/index.tsx [hunk 2]
```diff
@@ -0,0 +1,124 @@
+import { FormEvent, SyntheticEvent, useState } from 'react';
+import { Box, Button, Flex, FormControl, FormLabel, FormErrorMessage, Heading, Input, SimpleGrid, Image } from '@chakra-ui/react';
+import Layout from '../../components/Layout';
+import useSnackbar from '../../hooks/useSnackbar';
+import Link from 'next/link';
+
+const LoginPage = () => {
+  const [email, setEmail] = useState('');
+  const [isValidEmail, setIsValidEmail] = useState(true); // So that error does not pop up on first visit
```
Comment (MorningLit) [inline pages/login/index.tsx:9]: i'm sorry but please use react-hook-form 😭

File: pages/login/index.tsx (hunk 1 above)
Comment (MorningLit) [reply pages/login/index.tsx:64]: either move this padding to the outerest most Box or remove it

**Comment by MorningLit** [comment]
Where did you get the design from?

File: pages/login/index.tsx [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+  return (
+    <Box height='100vh' display='flex' justifyContent='center' alignItems='center' backgroundColor='#f6f6f6'>
+      <Flex width='full' alignContent='center' justifyContent='center'>
+        <SimpleGrid columns={[1, 1, 1, 2]} spacing={0}>
+          <Box display={['none', 'none', 'none', 'block']}>
+            <Image
+              src='https://foodbank.sg/wp-content/uploads/2020/07/IMG_9279-1024x768.jpg'
+              alt='Food Bank'
+              backgroundPosition='center'
+              height='100vh'
```
Comment (MorningLit) [inline pages/login/index.tsx:58]: height 100% can be used here

File: pages/login/index.tsx [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+                src='https://i.imgur.com/7izz9oQ.png'
+                alt='Food Bank'
+                backgroundPosition='center'
+                height='200px'
+                marginBottom={6}
+              />
+              <Heading>Welcome back!</Heading>
+              <Box textAlign='left'>
+                <FormControl isRequired isInvalid={!isValidEmail} mt={4} width={{ sm: '80vw', md: '80vw', lg: '500px' }}>
+                  <FormLabel htmlFor='email'>Email</FormLabel>
```
Comment (MorningLit) [inline pages/login/index.tsx:84]: image
title and labels are white on dark theme

File: pages/login/index.tsx [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+
+  const handlePasswordChange = (e: FormEvent<HTMLInputElement>) => {
+    setPassword(e.currentTarget.value);
+  };
+
+  const handleSubmit = async (e: SyntheticEvent) => {
+    e.preventDefault();
+    setIsLoading(true);
+    try {
+      const res = await fetch('/api/auth/login', {
```
Comment (kevinchua6) [inline pages/login/index.tsx:28]: Can you help me to change this to react query? Thanks!

File: pages/login/index.tsx [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+    setIsValidEmail(emailRegex.test(email));
+  };
+
+  const handlePasswordChange = (e: FormEvent<HTMLInputElement>) => {
+    setPassword(e.currentTarget.value);
+  };
+
+  const handleSubmit = async (e: SyntheticEvent) => {
+    e.preventDefault();
+    setIsLoading(true);
```
Comment (kevinchua6) [inline pages/login/index.tsx:26]: loading also can be replaced by react query!

File: pages/login/index.tsx [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+                // Placeholder url
+                src='https://i.imgur.com/7izz9oQ.png'
+                alt='Food Bank'
+                backgroundPosition='center'
+                height='200px'
+                marginBottom={6}
+              />
+              <Heading>Welcome back!</Heading>
+              <Box textAlign='left'>
+                <FormControl isRequired isInvalid={!isValidEmail} mt={4} width={{ sm: '80vw', md: '80vw', lg: '500px' }}>
```
Comment (kevinchua6) [inline pages/login/index.tsx:83]: Good job making this responsive!!

**Review by kevinchua6 — CHANGES_REQUESTED** [review]

File: pages/login/index.tsx (hunk 1 above)
Comment (kxrt) [reply pages/login/index.tsx]: Done!

File: pages/login/index.tsx (hunk 2 above)
Comment (kxrt) [reply pages/login/index.tsx]: Done!

File: pages/login/index.tsx (hunk 4 above)
Comment (kxrt) [reply pages/login/index.tsx]: Should be fixed now!

File: pages/login/index.tsx (hunk 7 above)
Comment (kxrt) [reply pages/login/index.tsx]: Thanks!

File: pages/login/index.tsx (hunk 5 above)
Comment (kxrt) [reply pages/login/index.tsx:21]: Should I be using Axios here too? I'm not sure of the API's url.

**Comment by kxrt** [comment]
> Where did you get the design from?

@MorningLit I searched up "split login page" for inspiration and free-hand edited from there.

File: pages/login/index.tsx (hunk 3 above)
Comment (kxrt) [reply pages/login/index.tsx]: Done!

File: pages/login/index.tsx (hunk 5 above)
Comment (kevinchua6) [reply pages/login/index.tsx:21]: yes please! yeah the api isn't done yet so just put a placeholder url in the meantime

Comment (kxrt) [reply pages/login/index.tsx]: Done!

File: pages/login/index.tsx (hunk 6 above)
Comment (kxrt) [reply pages/login/index.tsx]: Done!

File: pages/login/index.tsx [hunk 8]
```diff
@@ -2,53 +2,42 @@
...
-      });
-      const values = await res.json();
-      if (res.ok) {
-        // login successful
-        openSuccessNotification('Login successful', 'You are now logged in.');
-        window.location.href = '/';
-      }
-    } catch (err: any) {
-      console.error(err);
+  const mutation = useMutation(login, {
```
Comment (kevinchua6) [inline pages/login/index.tsx:27]: Good job :)

File: pages/login/index.tsx [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+                      borderColor='#78be20'
+                      {...register('email', {
+                        required: 'This is required.',
+                        pattern: {
+                          value: /^[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+@[a-zA-Z0-9-]+(?:\.[a-zA-Z0-9-]+)*$/,
+                          message: 'Please enter a valid email address.',
+                        },
+                      })}
+                    />
+                    {(errors.email || mutation.isError) && <FormErrorMessage>Please enter a valid email address.</FormErrorMessage>}
```
Comment (kevinchua6) [inline pages/login/index.tsx:80]: This is when the user enters a wrong username and password combination right? I feel that this error message is incorrect (an email is valid, just that it doesn't match the password). Maybe just remove this and put `Incorrect username or password. ` under the password field when the user fails to log in?

File: pages/login/index.tsx [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+  const { openErrorNotification, openSuccessNotification } = useSnackbar();
+
+  const login = (data: FormData) => {
+    return axios.post(url, data);
+  };
+
+  const mutation = useMutation(login, {
+    onSuccess: () => {
+      openSuccessNotification('Login successful', 'Welcome back!');
+      window.location.href = '/';
```
Comment (kevinchua6) [inline pages/login/index.tsx:30]: Use router.push instead, I think a native nextjs solution is better

File: pages/login/index.tsx [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+import { Box, Button, Flex, FormControl, FormLabel, FormErrorMessage, Heading, Input, SimpleGrid, Image } from '@chakra-ui/react';
+import Link from 'next/link';
+import { useForm } from 'react-hook-form';
+import axios from 'axios';
+import { useMutation } from '@tanstack/react-query';
+
+// Local imports
+import useSnackbar from '../../hooks/useSnackbar';
+
+const url = '/api/auth/login';
```
Comment (kevinchua6) [inline pages/login/index.tsx:12]: I think we should capitalise the variable name for this

File: pages/login/index.tsx [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+      openSuccessNotification('Login successful', 'Welcome back!');
+      window.location.href = '/';
+    },
+    onError: error => {
+      openErrorNotification('Login failed', 'Please recheck your email and password.');
+    },
+  });
+
+  const onSubmit = (data: FormData) => {
+    console.log(data);
```
Comment (kevinchua6) [inline pages/login/index.tsx:38]: Please remove this, thanks

**Review by kevinchua6 — CHANGES_REQUESTED** [review]
Looks good, just some small nits!

File: pages/login/index.tsx (hunk 12 above)
Comment (kxrt) [reply pages/login/index.tsx:38]: Done!

File: pages/login/index.tsx (hunk 11 above)
Comment (kxrt) [reply pages/login/index.tsx:12]: Makes sense, missed this out. Done!

File: pages/login/index.tsx (hunk 10 above)
Comment (kxrt) [reply pages/login/index.tsx:30]: Edited!

File: pages/login/index.tsx (hunk 8 above)
Comment (kxrt) [reply pages/login/index.tsx:27]: Thanks :)

File: pages/login/index.tsx (hunk 9 above)
Comment (kxrt) [reply pages/login/index.tsx]: Done with a slight modification - there's still an invalid email address error when it fails regex (through `errors.email`).

File: pages/login/index.tsx [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+import Link from 'next/link';
+import { useForm } from 'react-hook-form';
+import axios from 'axios';
+import { useMutation } from '@tanstack/react-query';
+import { useRouter } from 'next/router'
+
+// Local imports
+import useSnackbar from '../../hooks/useSnackbar';
+
+const URL = '/api/auth/login';
```
Comment (MorningLit) [inline pages/login/index.tsx:13]: move this to utils/ folder and create a link file to store all the URL links in one place

**Review by MorningLit — APPROVED** [review]
image
there's still some issue with the design 🙈 , but since this isn't the finalised UED design, lgtm!

File: pages/login/index.tsx (hunk 13 above)
Comment (kxrt) [reply pages/login/index.tsx]: Done!

**Review by kevinchua6 — APPROVED** [review]
LGTM, just merge master into it and it's fine

**Pull request merged**